src/main/java/com/cg_vibely_social_service/service/impl/ChatServiceImpl.java

```diff
@@ -24,18 +29,119 @@ public class ChatServiceImpl implements ChatService {
+    private final int MESSAGE_PAGE_SIZE = 20;
+    private final int MAX_CACHE_PAGES = 5;
+    private final int CACHED_MESSAGES_MAX_SIZE = 20;
 
     @Override
     public void save(ChatMessageDto messageDto) {
-        ChatMessage chatMessage = new ChatMessage();
-        BeanUtils.copyProperties(messageDto,chatMessage);
-        messageRepository.save(chatMessage);
+        String sender = messageDto.getSender();
+        String receiver = messageDto.getReceiver();
+        String liveChatCacheKey = String.format(MESSAGES_BYUSER_FORMAT, sender, receiver);
+        String rv_liveChatCacheKey = String.format(MESSAGES_BYUSER_FORMAT, receiver, sender);
+        try {
+            if (Boolean.TRUE.equals(redisTemplate.hasKey(liveChatCacheKey))) {
+                saveIfKeyExist(messageDto, sender, receiver, liveChatCacheKey);
+            } else if (Boolean.TRUE.equals(redisTemplate.hasKey(rv_liveChatCacheKey))) {
+                saveIfKeyExist(messageDto, sender, receiver, rv_liveChatCacheKey);
+            } else {
+                List<ChatMessageDto> chatMessageDtoList = new ArrayList<>();
+                chatMessageDtoList.add(messageDto);
+                ChatMessagesResponse chatMessagesResponse = new ChatMessagesResponse(1, chatMessageDtoList);
+                cachedData.set(liveChatCacheKey, gsonUtils.parseToString(chatMessagesResponse));
+            }
+        } catch (Exception e) {
+            ChatMessage chatMessage = new ChatMessage();
+            BeanUtils.copyProperties(messageDto, chatMessage);
+            messageRepository.save(chatMessage);
+        }
+    }
+
+    private void saveIfKeyExist(ChatMessageDto messageDto, String sender, String receiver, String liveChatCacheKey) {
+        String stringOfCachedCategories = cachedData.get(liveChatCacheKey);
+        List<ChatMessageDto> chatMessageDtoList = gsonUtils
+                .parseToObject(stringOfCachedCategories, ChatMessagesResponse.class)
+                .getMessageList();
+        chatMessageDtoList.add(messageDto);
+
+        if (chatMessageDtoList.size() >= CACHED_MESSAGES_MAX_SIZE) {
+            List<ChatMessage> chatMessages = chatMessageConverter.convert(chatMessageDtoList);
+            messageRepository.saveAll(chatMessages);
+
+            List<String> cachedKeys = new ArrayList<>();
+            for (int page = 0; page <= MAX_CACHE_PAGES; page++) {
+                cachedKeys.add(String.format(PAGE_MESSAGE_BYUSER_FORMAT, sender, receiver, page));
+            }
+            redisTemplate.delete(cachedKeys);
+            redisTemplate.delete(liveChatCacheKey);
+        } else {
+            ChatMessagesResponse chatMessagesResponse = new ChatMessagesResponse(1, chatMessageDtoList);
+            cachedData.set(liveChatCacheKey, gsonUtils.parseToString(chatMessagesResponse));
+        }
+    }
+
+    @Override
+    public ChatMessagesResponse findAllWithCacheActive(String senderEmail, String receiverEmail, Pageable pageable) {
```

Comment:
The block code is too long, and should split into many functions instead put them into one function

Reply:
Resolved